Geofencing/app/src/main/java/com/google/android/gms/location/sample/geofencing/MainActivity.java

```diff
@@ -26,406 +26,411 @@
+   */
+  private void showSnackbar(
+      final int mainTextStringId, final int actionStringId, View.OnClickListener listener) {
+    Snackbar.make(
+            findViewById(android.R.id.content),
+            getString(mainTextStringId),
+            Snackbar.LENGTH_INDEFINITE)
+        .setAction(getString(actionStringId), listener)
+        .show();
+  }
+
+  /** Returns true if geofences were added, otherwise false. */
+  private boolean getGeofencesAdded() {
+    return PreferenceManager.getDefaultSharedPreferences(this)
+        .getBoolean(Constants.GEOFENCES_ADDED_KEY, false);
+  }
+
+  /**
+   * Stores whether geofences were added ore removed in {@link SharedPreferences};
+   *
+   * @param added Whether geofences were added or removed.
+   */
+  private void updateGeofencesAdded(boolean added) {
+    PreferenceManager.getDefaultSharedPreferences(this)
+        .edit()
+        .putBoolean(Constants.GEOFENCES_ADDED_KEY, added)
+        .apply();
+  }
+
+  /** Performs the geofencing task that was pending until location permission was granted. */
+  private void performPendingGeofenceTask() {
+    if (mPendingGeofenceTask == PendingGeofenceTask.ADD) {
+      addGeofences();
+    } else if (mPendingGeofenceTask == PendingGeofenceTask.REMOVE) {
+      removeGeofences();
     }
-
-    /**
-     * Performs the geofencing task that was pending until location permission was granted.
-     */
-    private void performPendingGeofenceTask() {
-        if (mPendingGeofenceTask == PendingGeofenceTask.ADD) {
-            addGeofences();
-        } else if (mPendingGeofenceTask == PendingGeofenceTask.REMOVE) {
-            removeGeofences();
-        }
+  }
+
+  /** Return the current state of the permissions needed. */
+  private boolean checkPermissions() {
+    int permissionState =
+        ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
+    if (permissionState != PackageManager.PERMISSION_GRANTED) {
+      return false;
     }
-
-    /**
-     * Return the current state of the permissions needed.
-     */
-    private boolean checkPermissions() {
-        int permissionState = ActivityCompat.checkSelfPermission(this,
-                Manifest.permission.ACCESS_FINE_LOCATION);
-        return permissionState == PackageManager.PERMISSION_GRANTED;
+    permissionState =
+        ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+    return permissionState == PackageManager.PERMISSION_GRANTED;
+  }
+
+  private void requestPermissions() {
+    boolean shouldProvideRationale =
+        ActivityCompat.shouldShowRequestPermissionRationale(
+                this, Manifest.permission.ACCESS_FINE_LOCATION)
+            && ActivityCompat.shouldShowRequestPermissionRationale(
+                this, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+
+    // Provide an additional rationale to the user. This would happen if the user denied the
```

Comment:
It's actually not always like this. I would skip this comment

Reply:
> It's actually not always like this. I would skip this comment

Thanks, I'll remove it.